src/System.Management.Automation/engine/parser/tokenizer.cs

```diff
@@ -3260,41 +3294,100 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, char s
 
                     if (real)
                     {
-                        double d;
-                        if (Double.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out d))
+                        if (Double.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out double d))
                         {
                             // TryParse incorrectly return +0 when the result should be -0, so check for that case
                             if (d == 0.0 && strNum[0] == '-')
                             {
                                 d = -0.0;
                             }
-                            if (suffix == 'l' || suffix == 'L')
+
+                            if (suffix == NumberSuffixFlags.Long)
                             {
-                                result = ((long)Convert.ChangeType(d, typeof(long), CultureInfo.InvariantCulture)) * multiplier;
+                                result = Convert.ToInt64(d * multiplier);
+                            }
+                            else if (suffix == NumberSuffixFlags.Short)
+                            {
+                                result = Convert.ToInt16(d * multiplier);
+                            }
+                            else if (suffix.HasFlag(NumberSuffixFlags.Unsigned))
+                            {
+                                ulong testresult = Convert.ToUInt64(d * multiplier);
```

Comment:
All methods like `Convert.ToUInt64` internally test MaxValue and MinValue. We could expand the methods, do the checks explicitly and exclude exceptions at all.

Reply:
You'll have to elaborate a bit, I'm not sure quite what your intent is. :)

Reply:
Sample for a short above ( `result = Convert.ToInt16(d * multiplier);`):
```c#
var value = d * multiplier;

if (value < short.MinValue || value > short.MaxValue)
{
    result = null;
    return false;
}

result = (short)value;
return true;
...
```

Reply:
We could replace the `if`-s with `switch` to get more readable code.

Reply:
We can't do a direct cast, because that truncates the double instead of rounding it properly. But yes, if we do that and then call Int16.Parse it should work.

Reply:
Re if -> switch: there are parts of this that you can *sort of* get prettier code. If this were PS I could throw the entire thing in a gorgeous switch. But... It's not. 

So the switches are pretty limited in scope here, but I'll see what I can do.

Reply:
> We can't do a direct cast, 

You can look how CoreFX do this in Convert class for double
https://source.dot.net/#System.Private.CoreLib/shared/System/Convert.cs,938

Reply:
We can grap from CoreFX needed Convert.Int16/Int32/Int64 methods and put them in Utils.cs (without throw statement).

Reply:
Do we have a `Utils.cs` in this repo? Or am I making a new one?

Reply:
I think it is best place https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/Utils.cs

Reply:
As opposed to [PsUtils.cs](https://github.com/PowerShell/PowerShell/blob/5d03e1653a7d518715fa3f00587cad6b5c78cc89/src/System.Management.Automation/utils/PsUtils.cs)

Reply:
Thanks, guys! :D

Okay, got some helper functions down, trying to work on sensible ways to have the code structured.

Reply:
Well let's see how that goes...